### PR TITLE
[prover] fine-grained read and write set analysis

### PR DESCRIPTION
**Pull request by sblackshear**
### Overview

The analysis introduced in this PR seeks to overapproximate the set of *concrete access paths* that may be read/written by a given procedure. An access path is a canonical name for a location in memory; some examples of are:

`0x7/M::T/f` (i.e., the field `f` of the `M::T` resource stored at address `0x7`)

`Formal(0)/[2]` (i.e., the value stored at index 2 of the array bound the 0th formal of the current procedure)

The analysis uses an abstraction of access paths with the following components

- A *root*, which is either an abstract address or a local
- Zero or more *offsets*, where an offset is a field, an unknown vector index, or an abstract struct type

Abstract addresses are a set containing constants and abstract access paths read from the environment (or "footprint"). For example, in the following code:

```
struct S { f: u64 }

fun foo(x: S) {
  let a = if (*) { 0x1 } else { *&x.f }
   ...
}
```

we will represent `a` as `{ 0x1, Footprint(x/f) }`

### Access Path Trie

The obvious approach to abstracting a set of concrete paths is using a set of abstract paths. The analysis does exactly this, but using an implementation that avoids redundant representations of the same memory: an *access path trie* where roots nodes are access path roots and each internal node is an access path offset. Each node is (optionally) associated with abstract value of a generic type `T`.

The read/write set analysis state consists of two such tries: an `AccessPathTrie<Access>` to track the accesses to global state, and an `AccessPathTrie<AbsAddr>` to track the set of address values at each program point. An access is either Borrow, Read, Write, or ReadWriteBorrow (top value). The Borrow value is not strictly needed, but it helps for debugging.

### Read/Write Set Analysis

The read/write set analysis is a compositional analysis that starts from the leaves of the call graph and analyzes each procedure once. The result is a summary of the abstract paths read/written by each procedure and the value(s) returned by the procedure.

When the analysis encounters a call, it fetches the summary for the callee and applies it to the current state. This logic (implemented in `apply_summary`) is by far the most complex part of the analysis. There are three steps:

1. Substitute footprint values in the callee summary with their values in the caller state (including both actuals and values read from globals)
2. Bind return values in the callee summary to the return variables in the caller state
3. Join caller accesses and callee accesses

### Running

You can point the analysis at modules or transaction scripts via the prover frontend. For example, `mvp --read-write-set rotate_authentication_key.move` yields

```
Accesses:
Loc(0): Read
Loc(0)/0x1::DiemAccount::DiemAccount: Read
Loc(0)/0x1::DiemAccount::DiemAccount/key_rotation_capability: Borrow
Loc(0)/0x1::DiemAccount::DiemAccount/key_rotation_capability/vec: Read
Loc(0)/0x1::DiemAccount::DiemAccount/key_rotation_capability/vec/[_]: ReadWriteBorrow
Loc(0)/0x1::DiemAccount::DiemAccount/key_rotation_capability/vec/[_]/account_address: Read
Loc(0)/0x1::DiemAccount::DiemAccount/key_rotation_capability/vec/[_]/account_address/0x1::DiemAccount::DiemAccount: Read
Loc(0)/0x1::DiemAccount::DiemAccount/key_rotation_capability/vec/[_]/account_address/0x1::DiemAccount::DiemAccount/authentication_key: Write
Loc(0)/0x1::DiemAccount::DiemAccount/key_rotation_capability/vec/[_]/account_address/0x1::DiemAccount::DiemAccount/key_rotation_capability: Borrow
Loc(0)/0x1::DiemAccount::DiemAccount/key_rotation_capability/vec/[_]/account_address/0x1::DiemAccount::DiemAccount/key_rotation_capability/vec: Read
Loc(0)/0x1::DiemAccount::DiemAccount/key_rotation_capability/vec/[_]/account_address/0x1::DiemAccount::DiemAccount/key_rotation_capability/vec/[_]: ReadWriteBorrow
Loc(1): Read
```

### Future Work

There is a lot left to do before this analysis is useful/mature:
- Add many, many more tests. There are a few in this PR, and I have run on the Diem Framework/eyeballed several scripts to make sure the results look reasonable. But there are a ton of corner cases to test here.
- Expand the analysis from procedures to transactions: we must substitute in the concrete values provided in a transaction into the summary + (possibly) read some values from the current on-chain state. In addition, this analysis must take the prologue and epilogue into account.
- Make it possible to build a `GlobalEnv` from bytecode
- Address the termination issues by adding widening. Today, the analysis will hang if you write "address linked list" code that creates access paths of unknown length. We don't have any code like this in the framework, but it is important to handle this case in an open system
- Improve the precision of the analysis by adding guards to each access node. This would help us with the "all txes conflict because of gas issue", at least in the case where the price is nonzero.

### CODE REVIEW

**Comment by sblackshear** [comment]
@gdanezis

File: language/move-prover/bytecode/src/access_path.rs [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+        AbsStructTypeDisplay { s: self, env }
+    }
+}
+
+pub struct AbsStructTypeDisplay<'a> {
+    s: &'a AbsStructType,
+    env: &'a GlobalEnv,
+}
+
+impl<'a> fmt::Display for AbsStructTypeDisplay<'a> {
```
Comment (wrwg) [inline language/move-prover/bytecode/src/access_path.rs:89]: Nit: I would move the fmt::Display implementation to the end of the file. They are boilerplate distracting from understanding the data structures. Perhaps also all impl S after the data model has been introduced.

File: language/move-prover/bytecode/src/access_path.rs [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Addr {
+    /// Account address constant
+    Constant(Address),
+    /// Account address read from given access path
+    Footprint(AccessPath),
```
Comment (wrwg) [inline language/move-prover/bytecode/src/access_path.rs:113]: Is Footprint a technical term used for this kind of thing in program analysis?

File: language/move-prover/bytecode/src/access_path_trie.rs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+};
+use move_model::{model::FunctionEnv, ty::Type};
+use std::{
+    collections::btree_map::Entry,
+    fmt,
+    fmt::Formatter,
+    ops::{Deref, DerefMut},
+};
+
+/// The obvious approach to abstracting a set of concrete paths is using a set of abstract paths.
```
Comment (wrwg) [inline language/move-prover/bytecode/src/access_path_trie.rs:17]: Interesting question whether using a trie here instead of a flat set with redundancy is worth the complexity. I would expect 95% of the paths to be of length 2-4 (?). Or does the trie also helpful for the algorithms?

File: language/move-prover/bytecode/src/read_write_set_analysis.rs [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+                    let callee_fun_env = global_env.get_function(fun_id);
+                    if let Some(callee_summary) = self.cache.get::<ReadWriteSetState>(fun_id) {
+                        state.apply_summary(callee_summary, args, types, rets)
+                    } else {
+                        // native fun. use handwritten model
+                        match (
+                            callee_fun_env.module_env.get_identifier().as_str(),
+                            callee_fun_env.get_identifier().as_str(),
+                        ) {
+                            ("BCS", "to_bytes") => {
```
Comment (wrwg) [inline language/move-prover/bytecode/src/read_write_set_analysis.rs:374]: This should be factored out into its own function. Also we should likely add an enum somewhere centrally (TODO).

File: language/move-prover/bytecode/src/read_write_set_analysis.rs [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+        };
+        analysis.summarize(func_env, initial_state, data)
+    }
+
+    fn name(&self) -> String {
+        "read_write_set_analysis".to_string()
+    }
+}
+
+// ==== summary printing
```
Comment (wrwg) [inline language/move-prover/bytecode/src/read_write_set_analysis.rs:555]: Suggest to make those section markers more systematic and nicer. Also maybe use sections more extensively to organize this and other new code. Please also add regular doc comments more exhaustively.

File: language/move-prover/bytecode/src/read_write_set_analysis.rs [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+pub fn format_read_write_set_annotation(
+    target: &FunctionTarget<'_>,
+    code_offset: CodeOffset,
+) -> Option<String> {
+    // hack: the summary only contains the state at the exit block, but the
```
Comment (wrwg) [inline language/move-prover/bytecode/src/read_write_set_analysis.rs:560]: Good point. Perhaps add a TODO that we change the interface to allow both for function summary and per-bytecode printing (with both optional).

**Review by wrwg — COMMENTED** [review]
Looks very promising. Some initial comments.

File: language/move-prover/bytecode/src/access_path.rs (hunk 2 above)
Comment (sblackshear) [reply language/move-prover/bytecode/src/access_path.rs:113]: Yes. It is especially common in the separation logic community.

File: language/move-prover/bytecode/src/read_write_set_analysis.rs (hunk 4 above)
Comment (sblackshear) [reply language/move-prover/bytecode/src/read_write_set_analysis.rs:374]: Agree. Hopefully, most of these non-Novi specific native operations (e.g., vector stuff) will become bytecodes in time.

Comment (sblackshear) [reply language/move-prover/bytecode/src/read_write_set_analysis.rs]: Done

File: language/move-prover/bytecode/src/read_write_set_analysis.rs (hunk 5 above)
Comment (sblackshear) [reply language/move-prover/bytecode/src/read_write_set_analysis.rs]: Done

File: language/move-prover/bytecode/src/read_write_set_analysis.rs (hunk 6 above)
Comment (sblackshear) [reply language/move-prover/bytecode/src/read_write_set_analysis.rs:575]: Done

File: language/move-prover/bytecode/src/access_path.rs [hunk 7]
```diff
@@ -0,0 +1,744 @@
...
+    ast::TempIndex,
+    model::{GlobalEnv, ModuleId, QualifiedId, StructId},
+    ty::{PrimitiveType, Type, TypeDisplayContext},
+};
+use num::BigUint;
+use std::{fmt, fmt::Formatter};
+
+type Address = BigUint;
+
+/// This file contains an abstraction of concrete *access paths*, which are canonical names for a particular cell in memory. Some examples of concrete paths are:
```
Comment (wrwg) [inline language/move-prover/bytecode/src/access_path.rs:19]: nit: format

File: language/move-prover/bytecode/src/read_write_set_analysis.rs [hunk 8]
```diff
@@ -0,0 +1,645 @@
...
+};
+use move_model::{
+    ast::TempIndex,
+    model::{FunctionEnv, GlobalEnv, ModuleId, StructId},
+    ty::Type,
+};
+use std::{cmp::Ordering, fmt, fmt::Formatter};
+use vm::file_format::CodeOffset;
+
+/// The read/write set analysis is a compositional analysis that starts from the leaves of the
```
Comment (wrwg) [inline language/move-prover/bytecode/src/read_write_set_analysis.rs:21]: nit: I think you need to use `//!` for module level documentation comments. Technically, if not, this text is associated with the next item which is `enum Access`. My understanding is `//!` allows to associate doc with the enclosing item (the module).

**Review by wrwg — APPROVED** [review]
I think this is ready to land. I've not read all details, but we can iterate as it evolves.
